src/codegen/function_body_chunk.rs

```diff
@@ -119,15 +172,261 @@ impl Builder {
         self.add_in_array_lengths(&mut chunks);
         self.add_assertion(&mut chunks);
 
-        if let Some(ref trampoline) = self.async_trampoline {
+        if !self.callbacks.is_empty() || !self.destroys.is_empty() {
+            // Key: user data index
+            // Value: the current pos in the tuple for the given argument.
+            let mut poses = HashMap::with_capacity(group_by_user_data.len());
+            for trampoline in self.callbacks.iter() {
+                let user_data_index = trampoline.user_data_index;
+                let pos = poses.entry(user_data_index).or_insert_with(|| 0);
+                self.add_trampoline(env,
+                                    &mut chunks,
+                                    trampoline,
+                                    &group_by_user_data.get(&user_data_index).unwrap().1,
+                                    *pos,
+                                    &bounds,
+                                    &bounds_names,
+                                    false);
+                *pos += 1;
+            }
+            for destroy in self.destroys.iter() {
+                self.add_trampoline(env,
+                                    &mut chunks,
+                                    destroy,
+                                    &group_by_user_data.get(&destroy.user_data_index).unwrap().1,
+                                    0, // doesn't matter for destroy
+                                    &bounds,
+                                    &bounds_names,
+                                    true);
+            }
+            for (_, (pos, full_type, ref calls)) in group_by_user_data.iter() {
+                if calls.len() > 1 {
+                    chunks.push(
+                        Chunk::Let {
+                            name: format!("super_callback{}", pos),
```

Comment:
Why `super`? Above what?

Reply:
It's not `super` in the rust meaning but "super" as "wonderful". :)

Reply:
Maybe just call it `closure_X` as it's a closure and it's called like that by gir too?

Reply:
It's not a closure, strictly speaking. But If you have a suggestion for another name, please go ahead. The advantage of this one is that it's easy to grep into gir's source code.